### PR TITLE
Handle and propagate errors when checking if paths are Dirs, Files or Exist

### LINKED ISSUE #13175: Cron tasks can not find git repos time to time
- Gitea version (or commit ref): 1.12.4
- Git version: 2.28.0
- Operating system: FreeBSD 12.2 amd64
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [ ] Not relevant
- Log gist:

## Description
In notices from past 3 days we got:
```
Failed to health check repository (ZZZ/RRRR): chdir /var/db/gitea/repositories/zzz/rrrr.git: no such file or directory
Failed to health check repository (ZZZ/RRRR): exit status 4 - error in commit...
Failed to health check repository (ZZZ/RRRR): chdir /var/db/gitea/repositories/zzz/rrrr.git: no such file or directory
Failed to health check repository (ZZZ/RRRR): exit status 4 - error in commit...
Failed to health check repository (ZZZ/RRRR): exit status 128 - fatal: not a git repository (or any of the parent directories): .git\n
```
ZZZ/RRRR - has wrong commits and ```exit status 4``` mean that fsck run ok.
/var/db/gitea - symlink (```ln -s```) to some location, never changed.
On other mine installation I have same config, except no symlink and have no issues with cron+repos.

This bug + https://github.com/go-gitea/gitea/issues/13174 = gitea removes all my repos every 72 hours.

Possible this is FreeBSD specific, but it is very annoying to lost all repos record.
Before 1.12.4 we got only fsck errors.

### PR DESCRIPTION
**Pull request by zeripath**
`com.IsFile` and the like although simple do elide the difference with errors on filesystem reads and may cause incorrect intepretations.

This PR switches all uses of from `com.Is*` to `util.Is*` and makes a best guess as to what to do - mostly log and propagate the error up.

This may make some misbehaving systems and OSs appear less stable but ... at least the problems are not being elided anymore.

Fix #13175 

(I'm calling this a fix for that issue because although I do not know what is behind the underlying issue here - a new bug report with logs following on from this change would be more useful.)

### CODE REVIEW

File: modules/templates/static.go [hunk 1]
```diff
@@ -77,8 +78,11 @@ func NewTemplateFileSystem() templateFileSystem {
 	}
 
 	customDir := path.Join(setting.CustomPath, "templates")
-
-	if com.IsDir(customDir) {
+	isDir, err = util.IsDir(customDir)
```
Comment (zeripath) [inline modules/templates/static.go]:
```suggestion
	isDir, err := util.IsDir(customDir)
```

File: modules/templates/static.go [hunk 2]
```diff
@@ -170,8 +174,11 @@ func Mailer() (*texttmpl.Template, *template.Template) {
 	}
 
 	customDir := path.Join(setting.CustomPath, "templates", "mail")
-
-	if com.IsDir(customDir) {
+	isDir, err := util.IsDir(customDir)
+	if err != nil {
+		return log.Warn("Failed to check if custom directory %s is a directory. %v", err)
```
Comment (zeripath) [inline modules/templates/static.go]:
```suggestion
		log.Warn("Failed to check if custom directory %s is a directory. %v", err)
```

File: routers/user/setting/adopt.go [hunk 3]
```diff
@@ -30,10 +30,17 @@ func AdoptOrDeleteRepository(ctx *context.Context) {
...
+	if err != nil {
 		ctx.ServerError("IsRepositoryExist", err)
 		return
-	} else if has || !com.IsDir(filepath.Join(root, dir+".git")) {
+	}
+
+	isDir, err := util.IsDir(filepath.Join(root, dir+".git"))
+	if err != nil {
+		ctx.ServerError("IsDir", err)
+	}
```
Comment (rozhuk-im) [inline routers/user/setting/adopt.go:43]: ```return``` should be here, you can not continue on error

File: routers/admin/repos.go [hunk 4]
```diff
@@ -120,10 +120,16 @@ func AdoptOrDeleteRepository(ctx *context.Context) {
...
-	if has, err := models.IsRepositoryExist(ctxUser, repoName); err != nil {
+	has, err := models.IsRepositoryExist(ctxUser, repoName)
+	if err != nil {
 		ctx.ServerError("IsRepositoryExist", err)
 		return
-	} else if has || !com.IsDir(models.RepoPath(ctxUser.Name, repoName)) {
+	}
+	isDir, err := util.IsDir(models.RepoPath(ctxUser.Name, repoName))
+	if err != nil {
+		ctx.ServerError("IsDir", err)
```
Comment (zeripath) [inline routers/admin/repos.go:130]:
```suggestion
		ctx.ServerError("IsDir", err)
		return
```

File: routers/api/v1/admin/adopt.go [hunk 5]
```diff
@@ -92,10 +92,16 @@ func AdoptRepository(ctx *context.APIContext) {
...
-	if has, err := models.IsRepositoryExist(ctxUser, repoName); err != nil {
+	has, err := models.IsRepositoryExist(ctxUser, repoName)
+	if err != nil {
 		ctx.InternalServerError(err)
 		return
-	} else if has || !com.IsDir(models.RepoPath(ctxUser.Name, repoName)) {
+	}
+	isDir, err := util.IsDir(models.RepoPath(ctxUser.Name, repoName))
+	if err != nil {
+		ctx.InternalServerError(err)
```
Comment (zeripath) [inline routers/api/v1/admin/adopt.go:102]:
```suggestion
		ctx.InternalServerError(err)
		return
```

File: routers/api/v1/admin/adopt.go [hunk 6]
```diff
@@ -147,10 +153,16 @@ func DeleteUnadoptedRepository(ctx *context.APIContext) {
...
```
Comment (zeripath) [inline routers/api/v1/admin/adopt.go:164]:
```suggestion
		ctx.InternalServerError(err)
		return
```

File: routers/repo/repo.go [hunk 7]
```diff
@@ -351,8 +350,11 @@ func Download(ctx *context.Context) {
 		return
 	}
 	refName = strings.TrimSuffix(uri, ext)
-
-	if !com.IsDir(archivePath) {
+	isDir, err := util.IsDir(archivePath)
+	if err != nil {
+		ctx.ServerError("Download -> IsDir", err)
```
Comment (zeripath) [inline routers/repo/repo.go]:
```suggestion
		ctx.ServerError("Download -> IsDir", err)
		return
```

File: routers/user/setting/adopt.go [hunk 8]
```diff
@@ -30,10 +30,17 @@ func AdoptOrDeleteRepository(ctx *context.Context) {
...
+	has, err := models.IsRepositoryExist(ctxUser, dir)
+	if err != nil {
 		ctx.ServerError("IsRepositoryExist", err)
 		return
-	} else if has || !com.IsDir(filepath.Join(root, dir+".git")) {
+	}
+
+	isDir, err := util.IsDir(filepath.Join(root, dir+".git"))
+	if err != nil {
+		ctx.ServerError("IsDir", err)
```
Comment (zeripath) [inline routers/user/setting/adopt.go:41]:
```suggestion
		ctx.ServerError("IsDir", err)
		return
```

**Review by zeripath — COMMENTED** [review]
missed returns post ctx.ServerError

File: routers/user/setting/adopt.go (hunk 3 above)
Comment (zeripath) [reply routers/user/setting/adopt.go:43]: Thanks, it was a rather tiresome PR and I must have just forgotten to add these.

File: models/repo.go [hunk 9]
```diff
@@ -1524,11 +1556,15 @@ func updateRepository(e Engine, repo *Repository, visibilityChanged bool) (err e
 
 		// Create/Remove git-daemon-export-ok for git-daemon...
 		daemonExportFile := path.Join(repo.RepoPath(), `git-daemon-export-ok`)
-		if repo.IsPrivate && com.IsExist(daemonExportFile) {
+		isExist, err := util.IsExist(daemonExportFile)
+		if err != nil {
+			log.Error("Unable to check if %s exists. Error: %v", daemonExportFile, err)
```
Comment (rozhuk-im) [inline models/repo.go:1565]: ```return err```, in next code you check ```!isExist``` bu not check ```err```

File: modules/repository/hooks.go [hunk 10]
```diff
@@ -112,15 +111,27 @@ func CheckDelegateHooks(repoPath string) ([]string, error) {
 		newHookPath := filepath.Join(hookDir, hookName+".d", "gitea")
 
 		cont := false
-		if !com.IsExist(oldHookPath) {
+		isExist, err := util.IsExist(oldHookPath)
+		if err != nil {
+			results = append(results, fmt.Sprintf("unable to check if %s exists. Error: %v", oldHookPath, err))
+		}
+		if !isExist {
```
Comment (rozhuk-im) [inline modules/repository/hooks.go]: ```if err == nil && !isExist {```

File: modules/repository/hooks.go [hunk 11]
```diff
@@ -112,15 +111,27 @@ func CheckDelegateHooks(repoPath string) ([]string, error) {
...
 			results = append(results, fmt.Sprintf("old hook file %s does not exist", oldHookPath))
 			cont = true
 		}
-		if !com.IsExist(oldHookPath + ".d") {
+		isExist, err = util.IsExist(oldHookPath + ".d")
+		if err != nil {
+			results = append(results, fmt.Sprintf("unable to check if %s exists. Error: %v", oldHookPath+".d", err))
+		}
+		if !isExist {
```
Comment (rozhuk-im) [inline modules/repository/hooks.go]: ```if err == nil && !isExist {```

File: modules/repository/hooks.go [hunk 12]
```diff
@@ -112,15 +111,27 @@ func CheckDelegateHooks(repoPath string) ([]string, error) {
...
 			results = append(results, fmt.Sprintf("hooks directory %s does not exist", oldHookPath+".d"))
 			cont = true
 		}
-		if !com.IsExist(newHookPath) {
+		isExist, err = util.IsExist(newHookPath)
+		if err != nil {
+			results = append(results, fmt.Sprintf("unable to check if %s exists. Error: %v", newHookPath, err))
+		}
+		if !isExist {
```
Comment (rozhuk-im) [inline modules/repository/hooks.go]: ```if err == nil && !isExist {```

File: modules/repository/init.go [hunk 13]
```diff
@@ -192,7 +197,12 @@ func checkInitRepository(owner, name string) (err error) {
 }
 
 func adoptRepository(ctx models.DBContext, repoPath string, u *models.User, repo *models.Repository, opts models.CreateRepoOptions) (err error) {
-	if !com.IsExist(repoPath) {
+	isExist, err := util.IsExist(repoPath)
+	if err != nil {
+		log.Error("Unable to check if %s exists. Error: %v", repoPath, err)
+		return err
+	}
+	if !isExist {
```
Comment (rozhuk-im) [inline modules/repository/init.go:205]: ```if err == nil && !isExist {```

File: modules/ssh/ssh.go [hunk 14]
```diff
@@ -211,7 +212,11 @@ func Listen(host string, port int, ciphers []string, keyExchanges []string, macs
 	}
 
 	keyPath := filepath.Join(setting.AppDataPath, "ssh/gogs.rsa")
-	if !com.IsExist(keyPath) {
+	isExist, err := util.IsExist(keyPath)
+	if err != nil {
+		log.Fatal("Unable to check if %s exists. Error: %v", keyPath, err)
+	}
+	if !isExist {
```
Comment (rozhuk-im) [inline modules/ssh/ssh.go:219]: ```if err == nil && !isExist {```

File: routers/repo/repo.go [hunk 15]
```diff
@@ -389,7 +392,11 @@ func Download(ctx *context.Context) {
 	}
 
 	archivePath = path.Join(archivePath, base.ShortSha(commit.ID.String())+ext)
-	if !com.IsFile(archivePath) {
+	isFile, err := util.IsFile(archivePath)
+	if err != nil {
+		log.Error("Unable to check if %s is a file. Error: %v", archivePath, err)
+	}
+	if !isFile {
```
Comment (rozhuk-im) [inline routers/repo/repo.go]: ```if err == nil && !isExist {```

**Comment by rozhuk-im** [comment]
```if err == nil && !isExist {```
or ```else if !isExist {```
looks like I mark all suspect places.
In few places I found double error logging, but it is ok :)

File: modules/repository/hooks.go (hunk 10 above)
Comment (zeripath) [reply modules/repository/hooks.go]:
```suggestion
		if err == nil && !isExist {
```

File: modules/repository/hooks.go (hunk 11 above)
Comment (zeripath) [reply modules/repository/hooks.go]:
```suggestion
		if err == nil && !isExist {
```

File: modules/repository/hooks.go (hunk 12 above)
Comment (zeripath) [reply modules/repository/hooks.go]:
```suggestion
		if err == nil && !isExist {
```

File: modules/repository/init.go (hunk 13 above)
Comment (zeripath) [reply modules/repository/init.go:205]: Not necessary as we already return above this.

File: modules/ssh/ssh.go (hunk 14 above)
Comment (zeripath) [reply modules/ssh/ssh.go:219]: Not necessary as the fatal will kill server first

File: routers/repo/repo.go (hunk 15 above)
Comment (zeripath) [reply routers/repo/repo.go]: I considered this one carefully and decided that recreating the archive was the correct thing to do here.

File: models/repo.go (hunk 9 above)
Comment (zeripath) [reply models/repo.go:1565]: Hmm... The issue is not the recreation of the export file here - it's the non-deletion of the file is the repo has gone private. So on reflection I will return the error here.

```suggestion
			log.Error("Unable to check if %s exists. Error: %v", daemonExportFile, err)
			return err
```

File: modules/repository/hooks.go (hunk 10 above)
Comment (zeripath) [reply modules/repository/hooks.go]: This is small fry btw - getting multiple logs here is not a serious issue.

File: modules/repository/init.go (hunk 13 above)
Comment (rozhuk-im) [reply modules/repository/init.go:205]: sorry, I miss this

File: modules/ssh/ssh.go (hunk 14 above)
Comment (rozhuk-im) [reply modules/ssh/ssh.go:219]: If ```log.Fatal``` means: log + exit then ok.

**Comment by rozhuk-im** [comment]
IMHO in future other devs will make same errors, due to bad API design.
To prevent coding error better to change API to return not bool but enum: Exist, NotExist, Error and err.

**Comment by zeripath** [comment]
@rozhuk-im that is not idiomatic for go.

**Comment by rozhuk-im** [comment]
https://github.com/go-gitea/gitea/blob/f2a3abc683ad4b2177b7c7c6160a2c0b4316120a/modules/structs/task.go#L30
Probably, but simullar things are used.

**Review by techknowlogick — APPROVED** [review]

**Comment by 6543** [comment]
@zeripath can you resolve the conflicts?

**Comment by zeripath** [comment]
@6543 Conflicts resolved

**Review by 6543 — APPROVED** [review]

**Comment by lafriks** [comment]
Tests fail

**Comment by zeripath** [comment]
Agh ! it looks like when I resolved conflicts I introduced a bug in repo_form.go. Should now be fixed.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
🚀

**Comment by techknowlogick** [comment]
🚀

**Pull request merged**